controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -24,49 +27,74 @@
 import io.pravega.controller.stream.api.grpc.v1.Controller.TxnStatus;
+import io.pravega.controller.util.Config;
 import io.pravega.shared.protocol.netty.ConnectionFailedException;
 import io.pravega.shared.protocol.netty.FailingReplyProcessor;
 import io.pravega.shared.protocol.netty.PravegaNodeUri;
 import io.pravega.shared.protocol.netty.ReplyProcessor;
 import io.pravega.shared.protocol.netty.WireCommand;
 import io.pravega.shared.protocol.netty.WireCommandType;
 import io.pravega.shared.protocol.netty.WireCommands;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.LoggerFactory;
+
 import java.util.AbstractMap;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
-import org.apache.commons.lang3.tuple.ImmutablePair;
-import org.apache.commons.lang3.tuple.Pair;
-import org.slf4j.LoggerFactory;
 
 import static io.netty.buffer.Unpooled.wrappedBuffer;
-import static io.pravega.shared.segment.StreamSegmentNameUtils.getQualifiedStreamSegmentName;
-import static io.pravega.shared.segment.StreamSegmentNameUtils.getScopedStreamName;
-import static io.pravega.shared.segment.StreamSegmentNameUtils.getSegmentNumber;
-import static io.pravega.shared.segment.StreamSegmentNameUtils.getTransactionNameFromId;
+import static io.pravega.shared.segment.StreamSegmentNameUtils.*;
 
 public class SegmentHelper {
 
     private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(SegmentHelper.class));
 
     private final Supplier<Long> idGenerator = new AtomicLong(0)::incrementAndGet;
+    private final HostControllerStore hostStore;
+    private final AuthHelper authHelper;
+    // cache of connection manager for segment store nodes.
+    // Pravega Connection Manager maintains a pool of connection for a segment store and returns a connection from 
+    // the pool on the need basis. 
+    private final LoadingCache<PravegaNodeUri, SegmentHelperConnectionManager> cache;
+
+    public SegmentHelper(HostControllerStore hostControllerStore, ConnectionFactory clientCF, AuthHelper authHelper) {
+        this.hostStore = hostControllerStore;
+        this.authHelper = authHelper;
+        cache = CacheBuilder.newBuilder()
+                            .maximumSize(Config.HOST_STORE_CONTAINER_COUNT)
+                            // if a host is not accessed for 2 minutes, remove it from the cache
+                            .expireAfterAccess(5, TimeUnit.MINUTES)
+                            .removalListener((RemovalListener<PravegaNodeUri, SegmentHelperConnectionManager>) removalNotification -> {
+                                // Whenever a connection manager is evicted from the cache call shutdown on it. 
```

Comment:
Do we have guarentees that nobody else can have a reference to it at this point. IE: is there a race with git while this method is running etc. 
Please provide a comment pointing to guarentee if it exists.

Reply:
People can have reference to this class and can call getConnection and a new connection will get created. But when they return the connection, its not kept around but closed. And once all those that have references return connections, this could become a gc candidate.